### PR TITLE
[resharding] Introduce ReshardingManager

### PR DESCRIPTION
**Pull request by shreyan-gupta**
The resharding manager should hopefully be a one stop shop for all things resharding.

For now I'm keeping it as a member of Chain instead of Client but we can change that later.

The resharding manager in the future should be able to do the following work
- Figure out when to schedule a resharding event based on next epoch info, last block of epoch, shard layout etc.
- Manage all the sync preprocessing work like
  - Split parent trie into two parts
  - Update ShardTries with info about the two new child tries
  - Send request to any async job (probably we'll need a ReshardingActor)
- Manage all the sync postprocessing work like
  - Replace child tries with the newly created tries
  

There are some complications related to accessing store components that I plan to simplify in the future

### CODE REVIEW

**Comment by Trisfald** [comment]
Side note: I have something in an upcoming PR which might be more suited for resharding manager. It's about categorizing resharding events.

File: chain/chain/src/resharding/manager.rs [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+use near_primitives::types::AccountId;
+use near_primitives::utils::get_block_shard_id;
+use near_store::trie::mem::resharding::RetainMode;
+use near_store::{DBCol, PartialStorage, ShardTries, ShardUId, Store};
+
+pub struct ReshardingManager {
+    store: Store,
+    epoch_manager: Arc<dyn EpochManagerAdapter>,
+    /// Configuration for resharding.
+    pub resharding_config: MutableConfigValue<ReshardingConfig>,
```
Comment (wacban) [inline chain/chain/src/resharding/manager.rs:22]: Not for this PR; The old config won't make much sense for Instant Resharding. We can reuse the same config for it or introduce a separate config. I think we need to keep the old fields for now as we still use it for the archival nodes recovery.

File: chain/chain/src/resharding/manager.rs [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+    /// A handle that allows the main process to interrupt resharding if needed.
+    /// This typically happens when the main process is interrupted.
+    pub resharding_handle: ReshardingHandle,
```
Comment (wacban) [inline chain/chain/src/resharding/manager.rs:25]: FYI @Trisfald as we discussed interrupting resharding. This handle is how we did it in V2, it's very simple and it's already integrated so we can reuse that.

**Review by wacban — APPROVED** [review]
Looks great!

**Pull request merged**